src/Pos/Ssc/GodTossing/Seed.hs

```diff
@@ -64,19 +75,22 @@ calculateSeed (getCommitmentsMap -> commitments) openings lShares = do
     let mustBeRecovered :: HashSet StakeholderId
         mustBeRecovered = HS.difference participants (getKeys openings)
 
-    shares <- mapHelper BrokenShare (traverse (traverse fromBinaryM)) lShares
+    shares <- over _Left BrokenShare $
+              mapFailing (traverse (traverse fromBinaryM)) lShares
 
     -- Secrets recovered from actual share lists (but only those we need –
     -- i.e. ones which are in mustBeRecovered)
     let recovered :: HashMap StakeholderId (Maybe Secret)
+        -- NB. magical mystery list monad is used here
         recovered = HM.fromList $ do
             -- We are now trying to recover a secret for key 'k'
             id <- toList mustBeRecovered
             -- We collect all shares that 'k' has sent to other nodes
             let decryptedShares :: [Share]
-                decryptedShares = concatMap toList $ mapMaybe (HM.lookup id) (toList shares)
+                decryptedShares = concatMap toList $
```

Comment:
`concatMap` is `>>=` for list monad. So this piece of code can be rewritten usit `<-`